ABC Improvements

ismellike opened the pull request:
#697 Extension of this PR

- [X] Use a map for HATCHER_ALLOWLIST
- [X] Clean up state when going from Hatch -> Open
- [x] Various spell-checking fixes
- [x] Allow an existing tokenfactory token - reverted
- [x] Implement a circuit breaker
- [x] Add a new method to donate into the reserve [from comment](https://github.com/DA0-DA0/dao-contracts/pull/818#discussion_r1555016419) - reverted
- [x] Add transparency around initial token config - reverted
- [x] Fix logic around auto-sending funding pool [from comment](https://github.com/DA0-DA0/dao-contracts/pull/697#discussion_r1555019234)
- [x] Move curves into cw-curves package
- [x] Allow hatcher allowlist by DAO membership

ismellike committed:
Make hatcher_allowlist a map

Also cleans up state when going from Hatch -> Open
Various spell-checking fixes

ismellike committed:
Allow existing tokens to be used with ABC's

I think it's important for an existing token to be supported by an ABC.
Token DAO's will be able to set up liquidity more easily by simply allowing mint & burn functionality after instantiating the ABC while holding ownership of it.

*Also adds a query for hatcher allowlist
*fixes spell checks I ran into
*moves TokenInfo from dao-voting-token-staked to dao-interface for use in cw-abc

JakeHartnell commented on contracts/external/cw-abc/src/test_tube/integration_tests.rs:
```diff
@@ -514,3 +519,151 @@ fn test_update_curve() {
+}
+
+#[test]
+fn test_existing_token() {
```
For existing tokens, I think we need more tests to ensure no invariants. For example, if I use an existing token of which I have 1 million tokens, then the curve opens up and there are only enough funds to burn 500k... what happens if I try to sell the full amount? Just get zero for the rest of it?

One of the advantages of bonding curve based tokens is that tokens can't go to zero, but using existing tokens breaks this. Maybe not the end of the world, but we should think of test cases around this if we're going to support it.

JakeHartnell replied:
Also should be queries to know if there was a pre-mint, that's important context for users.

ismellike replied:
My understanding of this was that everyone could still sell, but people would start buying in deeper in the curve during the hatch phase. This would lead to less reserve relative to supply in the Open phase (compared to starting from 0 supply). 

This could be fine from the user-perspective if we introduce a new function for donation into the reserve instead of the funding pool. This would increase the reserve relative to supply and - for DAO's with profit being distributed back in - could fix this "ponzi-like" situation.  

Definitely a good idea to be able to show as much context as possible though. I will try to think of some specific test-cases as well.

ismellike replied:
After studying this some more, it's definitely not a good idea to have an existing token supply. DAO's wanting to use an ABC for liquidity should start as a membership-based DAO to instantiate the cw-abc and then update their voting module. I'm going to revert some of these changes here.

ismellike committed:
Complete funding pool logic

Renamed fees_recipient to funding_pool_forwarding for better naming
Made funding_pool_forwarding optional
Allow updating the funding_pool_forwarding by owner
Allow withdrawing from the funding pool by owner

ismellike committed:
Implement a circuit breaker

Adds IS_PAUSED state that is checked in execute entry
Also fixes naming in commands

ismellike committed:
Add query for initial supply at ABC creation

ismellike committed:
Allow donation into the reserve pool

Also fixes donation state being lost after multiple donations

ismellike committed:
Move cw-abc curves to its own package cw-curves

ismellike committed:
Improve validation for max_supply

ismellike committed:
Support DAO membership in hatcher allowlist

Could use a test here
Also fixes some warnings and flag for cw-std

ismellike committed:
Clippy fix

ismellike committed:
Only clone decimals from curve_state

also reorg buy command to be closer to sell command for readability

ismellike committed:
QoL improvements

Derive copy on DecimalPlaces
Catch OverflowError directly in ContractError
Do not set initial supply as supply - will further remove allowing initial supplies

ismellike committed:
Revert support for initial supply

ismellike committed:
Fix clippy unnecessary clones

ismellike committed:
Donate only into the funding pool

JakeHartnell commented on contracts/voting/dao-voting-token-staked/src/msg.rs:
```diff
@@ -3,6 +3,7 @@ use cosmwasm_std::{Binary, Uint128};
 use cw_utils::Duration;
 use dao_dao_macros::{active_query, native_token_query, voting_module_query};
 use dao_interface::token::NewTokenInfo;
+#[allow(unused_imports)]
```
Why were these needed?

ismellike replied:
This cleans up the build output, because it doesn't recognize the Response is actually used in the QueryResponses:
![image](https://github.com/DA0-DA0/dao-contracts/assets/5104272/1e12bdb4-4189-4f9b-aa57-fdd7f8d94631)

JakeHartnell commented on contracts/external/cw-abc/src/commands.rs:
```diff
@@ -283,32 +273,111 @@ pub fn execute_close(deps: DepsMut, info: MessageInfo) -> Result<Response, Contr
+}
+
+/// Withdraw funds from the funding pool (only callable by owner)
+pub fn withdraw(
```
Do we have any tests for this?

ismellike replied:
Adding some now

JakeHartnell reviewed (approved):
LFG!

ismellike committed:
Add a test for donate and withdraw from funding pool

JakeHartnell commented:
Running `just gen` will update the contract schemas and fix those linting errors.

ismellike committed:
Add test for dao hatchers + update schemas

Also removes unused deps from cw-abc which now live in cw-curves

ismellike committed:
Update Cargo.toml

ismellike commented:
Shoutout to @BHIIKTOR btw for noticing unnecessary packages and aliases after moving things to cw-curves package

ismellike commented:
I still want to squeeze in a couple of things.

- [x] Buy and sell quotes (how many tokens will a user receive from x amount)
- [x] MinMax contribution limit override by hatcher config

ismellike committed:
Buy & Sell Quotes

Let users know the curve state and returned amount after x payment

ismellike committed:
Complete DAO hatchers w/ priority queue

Disallow selling in the hatch phase
Do not validate max_contribution against max_raise

ismellike committed:
Fix priority queue ordering on ties

also clippy fixes

ismellike commented:
this is ready 🔥

ismellike committed:
Couple more clippy fixes
